gnovm/pkg/gnolang/preprocess.go

```diff
@@ -2805,6 +2805,200 @@ func findUndefined(store Store, last BlockNode, x Expr) (un Name) {
 	return findUndefined2(store, last, x, nil)
 }
 
+// finds the next undefined identifier and returns it if it is global
+func findUndefined2SkipLocals(store Store, last BlockNode, x Expr, t Type) Name {
+	name := findUndefined2(store, last, x, t)
+
+	existsLocal := func(name Name, bn BlockNode) bool {
+		curr := bn
+		for {
+			currNames := bn.GetBlockNames()
+
+			for _, currName := range currNames {
+				if currName == name {
+					return true
+				}
+			}
+
+			if bn.GetStaticBlock().Source == curr {
+				return false
+			}
+
+			curr = bn.GetStaticBlock().Source
+
+			if curr == nil {
+				return false
+			}
+		}
+	}
+
+	if name != "" {
+		pkg := packageOf(last)
+
+		if _, _, ok := pkg.FileSet.GetDeclForSafe(name); !ok {
+			// skip it if it's a local identifier
+			return ""
+		}
+
+		isLocal := existsLocal(name, last)
+
+		if isLocal {
+			return ""
+		}
+	}
+
+	return name
+}
+
+func findUndefinedStmt(store Store, last BlockNode, stmt Stmt, t Type) Name {
```

Comment:
Type expressions are generally not handled, both for ValueDecl, but also TypeDecl's. (Types can be declared within a function block).

This code panics, for instance:

```go
package main

var a = func() int {
	type B b
	x := B(1)
	return int(x)
}

type b int

func main() {
	println(a())
}
```